Added basic structure for e2e tests

insaaniManav opened the pull request.

insaaniManav commented on test/e2e_vet_test.go:
```diff
@@ -0,0 +1,19 @@
+		Dir: "scenarios/scripts",
+		Setup: func(env *testscript.Env) error {
+			// Set the environment variables
+			env.Setenv("E2E_VET_BINARY", "")
```
just as a sample here

insaaniManav commented:
@abhisek I have also added a sample of what a judo test file would look like , I think it looks cleaner , is easier to manage and is far more documented , would you mind evaluating this ?

abhisek commented on test/cli/vet-scans-vet.yml:
```diff
@@ -0,0 +1,6 @@
+run:
+  vetscansvet:                  # a "helloWorld" test suite "step"
+    command: 'echo $E2E_VET_BINARY scan -s --no-banner --lockfiles "$E2E_ROOT/go.mod" --report-summary=false --filter pkg.name == "github.com/safedep/dry"'
```
@insaaniManav The `echo` seems confusion here

insaaniManav commented:
@abhisek I have updated and added more tests to this , I could not replicate 5 and 6 . 6 seemed like a combination of all of 2 tests and 5 seems like it needs a supplementary shell script to verify the result of the test , judo can only interact with stderr or stdout I think all the CLI test frameworks only interact with them. 
For tests like 5 where we need to test the artifacts produced maybe we can run a supplementary shell script afterwards to verify all the artifacts.

abhisek commented:
> @abhisek I have updated and added more tests to this , I could not replicate 5 and 6 . 6 seemed like a combination of all of 2 tests and 5 seems like it needs a supplementary shell script to verify the result of the test , judo can only interact with stderr or stdout I think all the CLI test frameworks only interact with them. For tests like 5 where we need to test the artifacts produced maybe we can run a supplementary shell script afterwards to verify all the artifacts.

@insaaniManav While the YAML based approach looks clean and easy to get started, I think we will end up restricting ourselves in "expressing an E2E Scenario" if we go by it because we are unable to express 5 which is a valid and important scenario for vet's ability to build a dependency graph. I would use the following framework for decision making:

- Benefit: Does judo/YAML significantly simplify writing E2E scenarios over Bash/shell script that are generic and more expressive?
- Risk: Does adopting judo/testscript puts us in risk of wasting time trying to figure out how to express a scenario (e.g. 5) than what can be done with Bash?
- Reward: What are we gaining? Simplicity of E2E tests? Time saving? Better performance? Is it worth compared to risk?

Unless the reward is significantly larger than risk, I think we should put this on hold and continue with Bash for E2E scripts. This may not be pretty but gets the job done for now. It seems to me cli E2E testing is not easy and there is not much matured tooling around this space.

What do you think?

insaaniManav commented:
completely agreed the world needs more / a better e2e cli testing tool , we really would be restricting ourselves . its slightly cleaner but still not as powerful as bash will be
